src/Extension/AttributeExtension.php

```diff
@@ -0,0 +1,162 @@
+<?php
+
+/*
+ * This file is part of Twig.
+ *
+ * (c) Fabien Potencier
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Twig\Extension;
+
+use Twig\Attribute\AsTwigExtension;
+use Twig\Attribute\AsTwigFilter;
+use Twig\Attribute\AsTwigFunction;
+use Twig\Attribute\AsTwigTest;
+use Twig\Environment;
+use Twig\TwigFilter;
+use Twig\TwigFunction;
+use Twig\TwigTest;
+
+/**
+ * Define Twig filters, functions, and tests with PHP attributes.
+ *
+ * @author Jérôme Tamarelle <jerome@tamarelle.net>
+ *
+ * @internal
+ */
+final class AttributeExtension extends AbstractExtension
+{
+    private array $classes;
+    private array $filters;
+    private array $functions;
+    private array $tests;
+
+    /**
+     * A list of objects or class names defining filters, functions, and tests using PHP attributes.
+     * When passing a class name, it must be available in runtimes.
+     *
+     * @param class-string[]
+     */
+    public function __construct(array $classes)
+    {
+        $this->classes = $classes;
+    }
+
+    public function getFilters(): array
+    {
+        if (!isset($this->filters)) {
+            $this->initFromAttributes();
+        }
+
+        return $this->filters;
+    }
+
+    public function getFunctions(): array
+    {
+        if (!isset($this->functions)) {
+            $this->initFromAttributes();
+        }
+
+        return $this->functions;
+    }
+
+    public function getTests(): array
+    {
+        if (!isset($this->tests)) {
+            $this->initFromAttributes();
+        }
+
+        return $this->tests;
+    }
+
+    private function initFromAttributes()
+    {
+        $filters = $functions = $tests = [];
+
+        foreach ($this->classes as $objectOrClass) {
+            try {
+                $reflectionClass = new \ReflectionClass($objectOrClass);
+            } catch (\ReflectionException $e) {
+                throw new \LogicException(sprintf('"%s" class requires a list of objects or class name, "%s" given.', __CLASS__, get_debug_type($objectOrClass)), 0, $e);
+            }
+
+            $attributes = $reflectionClass->getAttributes(AsTwigExtension::class);
+            if (!$attributes) {
+                throw new \LogicException(sprintf('Extension class "%s" must have the attribute "%s" in order to use attributes', is_string($objectOrClass) ? $objectOrClass : get_debug_type($objectOrClass), AsTwigExtension::class));
+            }
+
+            foreach ($reflectionClass->getMethods() as $method) {
```

Comment:
thinking about that, I'm wondering whether those `asTwigFilter`, `AsTwigFunction` and `AsTwigTest` attributes should be read by that extension at all.

An alternative solution would be to have an extension in which you inject this list of test, filter and function metadata, and letting TwigBundle perform all this attribute reading at build-time instead. 
This would even removed the need for the `AsTwigExtension` attribute entirely (which would be quite confusing as the class having this attribute is _not_ a twig extension but a twig runtime) because the autoconfiguration system is able to apply autoconfiguration based on method-level attributes already.